feat: support svelte snippets

fedorovvvv opened the pull request:
Hi  🧡
I was using Svelte 5 and I have a need to insert icnoks like [snippet](https://svelte.dev/docs/svelte/snippet) 🚀

### Description

I solved this by adding a snippet to the component generation and exporting the snippet from it

The old API is not broken, but now can be used additionally:
```svelte
<script>
  import Icon, { snippet: iconSnippet } from '~icons/collection/name'
</script>

<Icon color='red'/>

{@render iconSnippet({
  color: 'red' 
})}
```
I think this is a very nice improvement for those using svelte 5 with runes 🧡

I also updated the d.ts file. For reasons I don't understand, Snippet doesn't want to accept props from SvelteHTMLElements if it is imported inside a declare module. Because of this, I took it out of it 🥲

### Linked Issues

I hope this helps a lot of people 😍

### Additional context

Yes, please validate the d.ts file of svelte 5.
This is the only option that works for me 🧡

userquin commented:
I'll check it tmr, I think the d.ts changes will break existing types

fedorovvvv commented:
> I'll check it tmr, I think the d.ts changes will break existing types

Thank you 🧡

If the option to import within module will work for you, let's do it this way.

I will be able to use my d.ts inside my project anyway and not use it from the library, but the main thing is not to break anything🥺

fedorovvvv commented:
And I also want to say that in svelte 5 `props` in regular `<Component/>` do not work.

This construction will not work and color will not be passed to svg
```svelte
<Icon color='red'/>
```

But with`snippet` everything works correctly
I noticed this a couple months ago. I should probably start an issue on this topic

fedorovvvv commented:
As I think it is possible to remove `{@html ...`
With and without it everything works properly, but `@html` can create a security hole

userquin commented:
Can you open an issue about the props issue? Do you have the types to your app.d.ts?

https://github.com/unplugin/unplugin-icons/blob/main/examples/sveltekit/src/app.d.ts#L1

userquin commented:
/cc @dominikg

userquin commented:
> As I think it is possible to remove `{@html ...`
> With and without it everything works properly, but `@html` can create a security hole

It is safe since the code will be replaced at build time, the content is just the compiled svg that comes from you local node_modulws, otherwise Sveltekit won't have that directive.

fedorovvvv commented:
> Can you open an issue about the props issue? Do you have the types in your app.d.ts?
> 
> https://github.com/unplugin/unplugin-icons/blob/main/examples/sveltekit/src/app.d.ts#L1
> 
> In the pr we test it and was working in VSCode and Jetbrains IDE (I tested IntelliJ and WebStorm).

Oh, thank you 🙏
It really does work wonders.

I will move the import to module in this PR

userquin commented:
This should be configurable, no idea yet about the perfomance  when using script context with the snippet per icon.

userquin commented on src/core/compilers/svelte.ts:
```diff
@@ -21,7 +21,7 @@ export const SvelteCompiler = (async (svg: string) => {
 
   sfc += svg.slice(closeTagStart)
-  return svelteRunes ? `<script>const{...p}=$props()</script>${sfc}` : sfc
+  return svelteRunes ? `<script module>export{snippet}</script><script>const{...p}=$props()</script>{#snippet snippet(p)}${sfc}{/snippet}{@render snippet(p)}` : sfc
```
Why do we need the script context here?

fedorovvvv replied:
According to the svelte documentation this is required to export a snippet from the .svelte component
https://svelte.dev/docs/svelte/snippet#Exporting-snippets

userquin replied:
Oh, I dond't see the export sorry

userquin commented:
If th3 snippet works better, we can just change the pr title to `use snippet intead {@html}` or somethibg similar.

fedorovvvv commented:
> This should be configurable, no idea yet about the perfomance when using script context with the snippet per icon.
> 
> You are not adding support, you are replacing the current implementation with the snippet.

We can render the component itself as before, without `{@render snippet}`.
Example: `<script module>export{snippet}</script><script>let p=$props()</script>{#snippet snippet(p)}${sfc}{/snippet}${sfc}`

fedorovvvv commented:
> If the snippet works better, we can just change the pr title to `feat: use snippet intead {@html}` or something similar.

I think @html is just a way to render svg code in dom
And snippet as one of the options to use the component itself. Both of these methods can exist together

fedorovvvv commented:
I think at this point it would make sense to use the `snippet` only for `export` and not use it for rendering in the component itself

Also, using `@html` was ~40% faster in load tests (10000 icons).

It's a wonderful way to go:
https://github.com/unplugin/unplugin-icons/pull/400#issuecomment-2764311251

fedorovvvv commented:
On the other hand, using a snippet within a component to export by default will not use the duplicated svg code, but will reuse from the snippet🥲

fedorovvvv commented:
It turned out that `@html ` doesn't matter at all in svelte runs because of the condition [code line](https://github.com/unplugin/unplugin-icons/pull/400/files#diff-9604cc434156d08714f75e10b42234d1f64535b91c86169e2f9d2af0db56207dR18), since it is not used in svelte 5+.

I suggest to switch completely to snippet to avoid duplicating svg code inside the component.

dominikg commented:
how would that work? The code emitted by the generated svelte5 icon components is pretty minimal already and compresses well.

And even if snippet mode performed slightly better, it looks much worse when adding icons 
now:
```html
<script>
import MyIcon from '...'
</script>
<MyIcon width={24}/>
```

with snippet:

```html
<script>
import {snippet as MyIconSnippet} from '...'
</script>
{@render MyIconSnippet({width: 24})
```

to avoid breaking changes for users, a "svelte-snippet" compiler could be added, but currently i still don't follow what kind of benefit it would bring. Depending on the app (amount of different icons, amount of same icon used in different places) it is possible that it has a benefit and i like the exploration here but my gut feeling right now is that snippet isn't the right tool here, at least not if its exposed to the enduser.

fedorovvvv commented:
Yeah, I've been thinking about a separate compiler too and am willing to make one

Snippet helps me to write components where I need to use "slots" 

For example:
```svelte
<Button startIcon={isonSnippet}>button<Button/>
```

In general, when the whole code is built on Snippets, it doesn't make much sense to use regular components for “slots”.

Of course, you can do it like this, but it doesn't make much sense to me because of the amount of additional code:

```svelte
<Button>
  {#snippet startIcon}
    <Icon/>
  {/startIcon}
  button
<Button/>
```

fedorovvvv commented:
Do I need to create additional examples and complete the documentation with svelte-snippet?

userquin commented:
The way you use unplugin-icons is not an issue with this repository, we can add a new compiler to allow use it and you'll be happy ;) , but current compilers are fine and this PR shouldn't break current Svelte compilers and the way consumers use the icons.

We can also add mix support for both in the same compiler (runes should be the default) adding a new mode (will require some change in the core), this way consumers will be able to use both current a snippet compilers.

I'm very busy, I'll try to send a PR based on this PR later if I have time, in the meantime you can give it a try.

userquin commented:
Add the new types in the package.json exports just below svelte5 here: https://github.com/unplugin/unplugin-icons/blob/main/package.json#L115-L117

fedorovvvv commented:
I've added 2 new expample
[sveltekit-snippet](https://github.com/fedorovvvv/unplugin-icons/tree/feature/svelte-snippets/examples/sveltekit-snippet)
[vite-svelte-snippet](https://github.com/fedorovvvv/unplugin-icons/tree/feature/svelte-snippets/examples/vite-svelte-snippet)

Made separate types for [svelte-snippet](https://github.com/fedorovvvv/unplugin-icons/blob/feature/svelte-snippets/types/svelte-snippet.d.ts) so that those using the regular svelte compiler don't have unnecessary types

Updated Build Tools in the readme:
[Readme](https://github.com/fedorovvvv/unplugin-icons/blob/feature/svelte-snippets/README.md#configuration)

Updated Frameworks in readme
[Readme](https://github.com/fedorovvvv/unplugin-icons/blob/feature/svelte-snippets/README.md#frameworks)

fedorovvvv commented:
> The way you use unplugin-icons is not an issue with this repository, we can add a new compiler to allow use it and you'll be happy ;) , but current compilers are fine and this PR shouldn't break current Svelte compilers and the way consumers use the icons.
> 
> We can also add mix support for both in the same compiler (runes should be the default) adding a new mode (will require some change in the core), this way consumers will be able to use both runes and snippet compilers.
> 
> I'm very busy, I'll try to send a PR based on this PR later if I have time, in the meantime you can give it a try.

snippet cannot be exported by default because the svlete 5 compiler does not allow it (

userquin commented:
Missing render example.

Remember that with this solution you can only use runes or snippet compiler, you cannot use both at the same time.

fedorovvvv commented:
> Missing render example.
> 
> Remember that with this solution you can only use runes or snippet compiler, you cannot use both at the same time.

Do I understand correctly that you mean Frameworks > Svelte?

Unfortunately svelte doesn't allow you to export snippets by default to use only them 😢

Right now the example looks like this:

```ts
imort Icon, {snippet as icon} from '~icons/collection/name'
```

dominikg commented:
in svelte5 you can pass in components as props and render them, so still not sure why you want/need a snippet from unplugin-icons itself

[repl](https://svelte.dev/playground/hello-world?version=5.25.3#H4sIAAAAAAAAA41Qy2rDMBD8la1acAIhplfXMbSn9hvqHvxYUxFZEtImaTH-964ch1qhgV4EOzua2ZlB6KpHkYlXVMrAyTjVwgpbSdiuxUZ0UqEX2fsg6NsGXgAYn389W7v1R1QUsLry-BfeGE2oiWVE7hsnLRWlLkkhQaDDDpLJN3kKsOytcQQvByKjoXOmh2SbnsdZMyK-NQtaGC6kwCl1nv5a6vzzsTgHHYLzeJenjDA-u3mqHAWN3RDeMeUdByD8IpGRO-C4uVFEdF_cxfXqRh0MewLLbTxYZ6xfra-PryehiTzcyw644su94xSW8gVU5Olymn6lspuYjZLNHnoMDrPq_2Iu-o1DxotlxFYeCwnWISMtkIEaodIgmcP1h21s_cFTJdVJ6lZkXaU8jj97hejCpQIAAA==)

dominikg commented:
i would recommend you join the svelte discord and create a thread in `#questions-svelte-and-kit` and ask for ideas regarding your dynamic icon usecase instead of working on this solution first. Thats usually the wrong way to tackle a problem.

fedorovvvv commented:
> in svelte5 you can pass in components as props and render them, so still not sure why you want/need a snippet from unplugin-icons itself
> 
> [repl](https://svelte.dev/playground/hello-world?version=5.25.3#H4sIAAAAAAAAA41Qy2rDMBD8la1acAIhplfXMbSn9hvqHvxYUxFZEtImaTH-964ch1qhgV4EOzua2ZlB6KpHkYlXVMrAyTjVwgpbSdiuxUZ0UqEX2fsg6NsGXgAYn389W7v1R1QUsLry-BfeGE2oiWVE7hsnLRWlLkkhQaDDDpLJN3kKsOytcQQvByKjoXOmh2SbnsdZMyK-NQtaGC6kwCl1nv5a6vzzsTgHHYLzeJenjDA-u3mqHAWN3RDeMeUdByD8IpGRO-C4uVFEdF_cxfXqRh0MewLLbTxYZ6xfra-PryehiTzcyw644su94xSW8gVU5Olymn6lspuYjZLNHnoMDrPq_2Iu-o1DxotlxFYeCwnWISMtkIEaodIgmcP1h21s_cFTJdVJ6lZkXaU8jj97hejCpQIAAA==)
> 
> dynamic components can also be typed https://svelte.dev/docs/svelte/typescript#The-Component-type so you should be able to restrict the type to an unplugin icon too

But what if my `Button` takes a Snippet and renders it as` {@render snippet()}`?

Or is this the right way to do it?
```svelte
<Button>
  {#snippet startIcon}
    <Icon/>
  {/startIcon}
  button
<Button/>
```
